docs/design/dft.md

```diff
@@ -0,0 +1,1006 @@
+# Mizar: Distributed Flow Table and Scaled Endpoint Design
+
+This is a strawman design for the Distributed Flow Table and Scaled
+Endpoint in which we detail its essential design techniques:
+
+-   Modular middleboxes with XDP programs that are chained to form
+  various types of network service decisions
+
+-   Local shared configuration across multiple XDP programs that reduces
+  the need for a packet to traverse multiple nodes to make a
+  decision on a new flow
+
+-   Overlay protocol integrated within the data-plane to provide
+  consistent connection state
+
+-   Built-in multi-tenancy that allows infrastructure sharing of network
+  services and middleboxes without the need to isolate middleboxes in
+  virtual machines or containers, hence improves cost
+
+-   Native data-plane failure detection that minimizes the time to
+  recover a connection by means of gossip protocols
+
+# Software Organization
+
+We will manage two separate pools of nodes. The
+first pool consists of data-plane nodes (D-Nodes) that are responsible
+for all packet processing. The other pool consists of management-plane
+nodes (M-Nodes) that run a set of Kubernetes custom resource operators
+and Kubernetes services. The operators implement the management-plane logic,
+and the other services cooperate with the operators to manage the
+D-Nodes and the middleboxes as we will detail in the Management Plane
+Section. The D,M-Nodes may join and leave the cluster dynamically,
+and the D-Nodes are not necessarily Kubernetes worker nodes and can
+be external to a Kubernetes cluster.
+
+The management-plane exposes network services as endpoints of a single
+IP address (VIP), each implements a chain of one or more middleboxes.
+We refer to such network service as (Scaled Endpoint), as an endpoint
+scale-out and in across multiple D-Nodes according to its bandwidth
+demands. The following is a non-exhaustive list of potential Scaled
+Endpoint types: Bouncer, Divider, Load-balancer, Nat, DPI, QoS,
+Tunnel, Firewall, Protocol Translators IPv6/IPv4, Access Control List
+(ACL) or a combination of these. A typical example from production
```

Comment:
If it is possible that we could provide a mechnism that let tenant users can choose their custom developed network service software as m-nodes? Or provide a open m-node framework, which tenant users can base on to develop their own m-nodes?

Reply:
I think this is a very good suggestion, we can certainly extend the m-node functionalities to allow tenants (in addition to admins) to define services themselves using basic building blocks (m-xdps) each doing a certain function. Currently, we intended to give this option to admins, but it's very possible to also extend it to users.